src/cudadecoder/cuda-fst.cc

```diff
@@ -106,27 +107,27 @@ void CudaFst::ComputeOffsets(const fst::StdFst &fst) {
 }
 
 void CudaFst::AllocateData(const fst::StdFst &fst) {
-  d_e_offsets_.reset(static_cast<uint32*>(CuDevice::Instantiate().Malloc(
-      (num_states_ + 1) * sizeof(*d_e_offsets_))));
-  d_ne_offsets_.reset(static_cast<uint32*>(CuDevice::Instantiate().Malloc(
-      (num_states_ + 1) * sizeof(*d_ne_offsets_))));
-  d_final_.reset(static_cast<float*>(CuDevice::Instantiate().Malloc(
-      (num_states_) * sizeof(*d_final_))));
+  void *temp_pointer;
+  CU_SAFE_CALL(cudaMalloc(&temp_pointer, (num_states_ + 1) * sizeof(uint32)));
```

Comment:
Replacing `Malloc` with `cudaMalloc` will not use the cache provided by `CuMemoryAllocator`. Will it have any side effects with such a replacement?

Reply:
This allocation happens only in constructors and destructors, so there will be no performance problems. The only possible effect may be a failure to allocate, if the custom memory allocator owns too much memory. See "This could break users who have given >99% of the the
available memory to the global allocator, though." in my initial comment. You could solve this by decreasing this value: https://github.com/kaldi-asr/kaldi/blob/e6cc1eef1f6c55936d36e8febe9baff6c479f81b/src/cudamatrix/cu-allocator.h#L74

Reply:
I do not remember what happens when the initial default allocation of 50% runs out? Will the allocator grab all available memory, or some part of it?

Besides, what is wrong with the allocator? I certainly don't like what it does with per-thread instantiation (I had to hack it to explicitly uninitialize every thread to use cuda-memcheck; even when I disabled it, it held handles to cudaBlas and other stuff). Once in a while I'm thinking of reworking its per-thread TLS use and syntax (this `CuDevice::Instantiate().Malloc` everywhere is ugly), but never finding a good reason. Or excuse. 

Here it's really unimportant, CudaFst constructor takes ungodly CPU time (2-3 seconds in a well-optimized build). But the allocator is used in other places too, I think. Are you planning to replace it everywhere in cudadecoder/?

Reply:
> Besides, what is wrong with the allocator?

Basically, the allocator prevents us from using the low level parts of cudadecoder codebase as a library for external codebases. When I say low-level, I mean cuda-fst.cc, cuda-decoder.cc, cuda-decoder-kernels.cu, and 
lattice-postprocessor.cc. It prevents us from doing so because it (Or CuDevice, rather than CuAllocator) makes a hard assumption that there is only one GPU per process. This means that we could not run the cudadecoder in pytorch when it is using multiple GPUs, for example.

There is another, smaller problem caused by the allocator when using this as a standalone library: Memory fragmentation is more likely. The user could set "cuda-memory-proportion" too high, which makes that memory unavailable to the rest of your application. There isn't a straightforward, easy way to check for this sort of error, but it can cause a performance degredation. By using cudaMalloc(), we allocate only that memory which is necessary.

These are the only locations in the cuda-fst.cc, cuda-decoder.cc, cuda-decoder-kernels.cu, and 
lattice-postprocessor.cc that use the cuda allocator, so I am basically allowing this part of the codebase to become a standalone library. Note that we don't need to remove usage of the allocator from the nnet3-pipeline*.cc files. Those already have strong dependencies on other kaldi parts (like nnet3), so there is no real benefit from doing so.

In general, the slowdown caused by cudaMalloc (it is a synchronous call, meaning that the call won't return until all previous enqueue operations and it complete) is not a huge problem when it is called within the constructor of an object (as is the case for CudaFst), since this happens only once. A caching allocator is useful in cases like deep neural network training where you have to allocate different sized chunks in each layer and don't want to suffer the synchronous stop caused by cudaMalloc().

Reply:
I'm OK with the way this is done using cudaMalloc.

Reply:
I see. The allocator is not used anywhere in cudadecoder/ except these places. Then it's certainly alright. And what about the cudafeat/ library?

Reply:
cudafeat/ library does use the caching allocator outside of constructors. See here: https://github.com/kaldi-asr/kaldi/blob/3b8a97d859464912ddc50877f7280048f654c275/src/cudafeat/online-ivector-feature-cuda.cc#L300

Not only that, there may be transitive dependencies on CuAllocator because of the usage of CuMatrix in that directory. So it would certainly be more complicated to "librarize" the cuda feature extractor software.

Reply:
@galv,
> cudafeat/ library does use the caching allocator outside of constructors. [...] Not only that...

So what do you think would work the best for you? It's hard to build a decoder without this library: the only option is extracting features on the host. Is my understanding correct: the only _blocker-level_ problem is our lack of support for multiple devices in one process? It would be cool to retain its speed where allocations come and go, and make it working for you at the same time. I already said I'm looking for an excuse to rework it a bit... :) 

You mentioned fragmentation, and that's fair, but show me an efficient allocator not affected by it. I productized Kaldi a few times (with and without CUDA). On Windows, I got bitten by the new allocator in VS 2013, which minimized fragmentation at a huge perf hit (I switched to TBBmalloc), and the GNU libc one, which suffers from memory creep to OOM in a few hours, and large perf hit with tunables set to prevent it (I switched to Jemalloc). BTW, is cudaMalloc immune to fragmentation, i.e., fully virtualized in hardware like mmap(3), but with a smaller equivalent of the "page size," the minimum allocation granularity?

By the way, I'm wondering why doesn't CUDA offer a suballocator like ours out of the box. I read TensorFlow code, and they RYO a very similar one. I'll be surprised if Torch had not have one as well.